synapse/storage/client_ips.py

```diff
@@ -90,6 +90,8 @@ def get_last_client_ip_by_device(self, devices):
             are (user_id, device_id) tuples. The values are also dicts, with
             keys giving the column names
         """
+        if not devices:
```

Comment:
the docstring says `devices` is an iterable rather than a list, and indeed there are call-sites where it is a generator. An empty generator is truthy, so this isn't going to work.

My inclination would be to skip this and instead bail if `where_clauses` ends up empty.